packages/react-reconciler/src/ReactFiberCompleteWork.new.js

```diff
@@ -966,6 +972,53 @@ function completeWork(
       bubbleProperties(workInProgress);
       return null;
     }
+    case Profiler: {
+      const didBailout = bubbleProperties(workInProgress);
+      if (!didBailout) {
+        // Use subtreeFlags to determine which commit callbacks should fire.
+        // TODO: Move this logic to the commit phase, since we already check if
+        // a fiber's subtree contains effects. Refactor the commit phase's
+        // depth-first traversal so that we can put work tag-specific logic
+        // before or after committing a subtree's effects.
+        const OnRenderFlag = Update;
+        const OnCommitFlag = Callback;
+        const OnPostCommitFlag = Passive;
+        const subtreeFlags = workInProgress.subtreeFlags;
+        const flags = workInProgress.flags;
+        let newFlags = flags;
+
+        // Call onRender any time this fiber or its subtree are worked on, even
+        // if there are no effects
+        newFlags |= OnRenderFlag;
+
+        // Call onCommit only if the subtree contains layout work, or if it
+        // contains deletions, since those might result in unmount work, which
+        // we include in the same measure.
+        // TODO: Can optimize by using a static flag to track whether a tree
+        // contains layout effects, like we do for passive effects.
```

Comment:
Can we though? The static flag wouldn't tell us if the layout effect happened to fire during the current update.

Reply:
Right but it would at least let us optimize the case where there are no layout unmount work in the tree at all. Probably not that big of a win, though, since most trees have at least one, like a DOM ref or something.